File: frontend/src/api/customers.ts
import api from './client'
import type { Customer, CustomerDetails, CustomerStats, Enquiry, EnquiryStats } from '@/types/customer'
import type { PaginatedResponse } from '@/types/auth'

// ── Customers ─────────────────────────────────────────────────────────────────
export const customersApi = {
  stats: () =>
    api.get<CustomerStats>('/customers/stats').then(r => r.data),

  list: (params?: {
    page?: number; page_size?: number; search?: string
    status?: string; customer_type?: string; kyc_status?: string
    assigned_user_id?: number; branch_id?: number; tag?: string
  }) =>
    api.get<PaginatedResponse<Customer>>('/customers', { params }).then(r => r.data),

  get: (id: number) =>
    api.get<Customer>(`/customers/${id}`).then(r => r.data),

  create: (data: Partial<Customer> & { phone: string; display_name: string }) =>
    api.post<Customer>('/customers', data).then(r => r.data),

  update: (id: number, data: Partial<Customer>) =>
    api.patch<Customer>(`/customers/${id}`, data).then(r => r.data),

  delete: (id: number) =>
    api.delete(`/customers/${id}`).then(r => r.data),

  upsertDetails: (id: number, data: Partial<CustomerDetails>) =>
    api.put<CustomerDetails>(`/customers/${id}/details`, data).then(r => r.data),

  updateKyc: (id: number, kyc_status: string) =>
    api.patch<Customer>(`/customers/${id}/kyc`, null, { params: { kyc_status } }).then(r => r.data),
}

// ── Enquiries ─────────────────────────────────────────────────────────────────
export const enquiriesApi = {
  stats: () =>
    api.get<EnquiryStats>('/enquiries/stats').then(r => r.data),

  list: (params?: {
    page?: number; page_size?: number; search?: string
    status?: string; source?: string; assigned_to_user_id?: number; is_converted?: boolean
  }) =>
    api.get<PaginatedResponse<Enquiry>>('/enquiries', { params }).then(r => r.data),

  get: (id: number) =>
    api.get<Enquiry>(`/enquiries/${id}`).then(r => r.data),

  create: (data: Partial<Enquiry> & { phone: string; full_name: string }) =>
    api.post<Enquiry>('/enquiries', data).then(r => r.data),

  update: (id: number, data: Partial<Enquiry>) =>
    api.patch<Enquiry>(`/enquiries/${id}`, data).then(r => r.data),

  delete: (id: number) =>
    api.delete(`/enquiries/${id}`).then(r => r.data),

  convert: (id: number, data: {
    customer_type?: string; display_name?: string
    pan?: string; gstin?: string; onboarded_at?: string
    assigned_user_id?: number; branch_id?: number
  }) =>
    api.post<Customer>(`/enquiries/${id}/convert`, data).then(r => r.data),
}

File: backend/app/api/v1/endpoints/customers.py
from fastapi import APIRouter, Depends, Query, Path
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
import math

from app.db.session import get_db
from app.core.deps import get_current_active_user
from app.models.auth import User
from app.schemas.customer import (
    CustomerCreate, CustomerUpdate, CustomerOut, CustomerListOut,
    CustomerDetailsCreate, CustomerDetailsOut,
    PaginatedResponse, MessageResponse,
)
from app.services import customer as svc

router = APIRouter(prefix="/customers", tags=["Customers"])


@router.get("/stats")
async def customer_stats(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    return await svc.get_customer_stats(db, current_user.tenant_id)


@router.get("", response_model=PaginatedResponse)
async def list_customers(
    page:             int           = Query(1, ge=1),
    page_size:        int           = Query(20, ge=1, le=100),
    search:           Optional[str] = Query(None),
    status:           Optional[str] = Query(None),
    customer_type:    Optional[str] = Query(None),
    kyc_status:       Optional[str] = Query(None),
    assigned_user_id: Optional[int] = Query(None),
    branch_id:        Optional[int] = Query(None),
    tag:              Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    customers, total = await svc.get_customers(
        db, current_user.tenant_id, page, page_size,
        search, status, customer_type, kyc_status,
        assigned_user_id, branch_id, tag,
    )
    return PaginatedResponse(
        items=[CustomerListOut.model_validate(c) for c in customers],
        total=total,
        page=page,
        page_size=page_size,
        total_pages=math.ceil(total / page_size) if total else 1,
    )


@router.post("", response_model=CustomerOut, status_code=201)
async def create_customer(
    data: CustomerCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    try:
        customer = await svc.create_customer(db, current_user.tenant_id, data)
        return CustomerOut.model_validate(customer)
    except Exception as e:
        import traceback
        print(f"Error creating customer: {e}")
        print(traceback.format_exc())
        raise


@router.get("/{customer_id}", response_model=CustomerOut)
async def get_customer(
    customer_id: int = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    customer = await svc.get_customer(db, current_user.tenant_id, customer_id)
    return CustomerOut.model_validate(customer)


@router.patch("/{customer_id}", response_model=CustomerOut)
async def update_customer(
    data: CustomerUpdate,
    customer_id: int = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    customer = await svc.update_customer(db, current_user.tenant_id, customer_id, data)
    return CustomerOut.model_validate(customer)


@router.delete("/{customer_id}", response_model=MessageResponse)
async def delete_customer(
    customer_id: int = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    await svc.delete_customer(db, current_user.tenant_id, customer_id)
    return MessageResponse(message="Customer deleted.")


# ─── Customer Details ─────────────────────────────────────────────────────────

@router.put("/{customer_id}/details", response_model=CustomerDetailsOut)
async def upsert_customer_details(
    data: CustomerDetailsCreate,
    customer_id: int = Path(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    details = await svc.update_customer_details(db, current_user.tenant_id, customer_id, data)
    return CustomerDetailsOut.model_validate(details)


# ─── KYC ─────────────────────────────────────────────────────────────────────

@router.patch("/{customer_id}/kyc", response_model=CustomerOut)
async def update_kyc(
    customer_id: int = Path(...),
    kyc_status: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    customer = await svc.update_kyc_status(db, current_user.tenant_id, customer_id, kyc_status)
    return CustomerOut.model_validate(customer)
